kvs/src/kvs/node_join_handler.cpp

```diff
@@ -86,7 +90,17 @@ void node_join_handler(
             kSelfTierIdVector, succeed, seed);
 
         if (succeed) {
-          if (threads.find(wt) == threads.end()) {
+          // there are two situations in which we gossip data to the joining
+          // node:
+          // 1) if the node is a new node and I am no longer responsible for
+          // the key
+          // 2) if the node is rejoining the cluster
+          // NOTE: This is currently inefficient because every server will
+          // gossip the key currently -- we might be able to hack around the
+          // has ring to do it more efficiently, but I'm leaving this here for
+          // now
+          if ((join_count == 0 && threads.find(wt) == threads.end()) ||
+              join_count > 0) {
```

Comment:
Seems like currently all threads gossip _all_ keys to the new node if its `join_count` > 0, but shouldn't we first check if the rejoining node/threads are responsible for the key to be more efficient?
Also, we probably shouldn't put the key into `join_remove_set` for the rejoining case, as the keys in this set will eventually be deleted, which we may not want.